spec/abstract/SchedContext_A.thy

```diff
@@ -195,72 +203,67 @@ where
-                \<comment> \<open>if refills has length at most @{text \<open>sc_refills_max sc\<close>}, then popping the head
-                    will ensure the refills are not full, so we may use False here\<close>
-         else do remnant \<leftarrow> return $ r_amount (hd refills) - usage;
-                 if remnant < MIN_BUDGET
-                 then if tl refills = []
-                      then set_refills sc_ptr [\<lparr>r_time = last_entry + period - remnant,
-                                                r_amount = r_amount (hd refills)\<rparr>]
-                      else do
-                        new_snd \<leftarrow> return $ (\<lparr>r_time = r_time (hd (tl refills)) - remnant,
-                                              r_amount = r_amount (hd (tl refills)) + remnant \<rparr>);
-                        set_refills sc_ptr (schedule_used False (new_snd # tl (tl refills)) used)
-                      od
-                 else do
-                   full \<leftarrow> refill_full sc_ptr;
-                   rfhd \<leftarrow> return $ (hd refills);
-                   new_head \<leftarrow> return $ (\<lparr>r_time = r_time rfhd + usage, r_amount = remnant\<rparr>);
-                   set_refills sc_ptr (schedule_used full (new_head # (tl refills)) used)
-                 od
-              od
+    usage' \<leftarrow> return $ min usage (r_amount (hd refills));
+
+    if usage' > 0
+    then do used \<leftarrow> return \<lparr>r_time = r_time (hd refills) + period, r_amount = usage'\<rparr>;
+            full \<leftarrow> refill_full sc_ptr;
+            refills' \<leftarrow> return $ schedule_used full refills used;
+            adjusted_hd \<leftarrow> return \<lparr>r_time = r_time (hd refills') + usage',
+                                   r_amount = r_amount (hd refills') - usage'\<rparr>;
+            set_refills sc_ptr (refills_merge_prefix (MIN_BUDGET_merge (adjusted_hd # (tl refills'))))
+         od
+    else set_refills sc_ptr (refills_merge_prefix (MIN_BUDGET_merge refills))
+
    od"
 
 definition
   refill_update :: "obj_ref \<Rightarrow> ticks \<Rightarrow> ticks \<Rightarrow> nat \<Rightarrow> (unit, 'z::state_ext) s_monad"
 where
   "refill_update sc_ptr new_period new_budget new_max_refills = do
+
+     set_sc_obj_ref sc_budget_update sc_ptr new_budget;
+     set_sc_obj_ref sc_period_update sc_ptr new_period;
+     set_sc_obj_ref sc_refill_max_update sc_ptr new_max_refills;
+
      sc \<leftarrow> get_sched_context sc_ptr;
-     refill_hd \<leftarrow> return $ refill_hd sc;
-     cur_time \<leftarrow> gets cur_time;
-     ready \<leftarrow> get_sc_refill_ready sc_ptr;
-     new_time \<leftarrow> return $ if ready then cur_time else (r_time refill_hd);
-     refill_hd \<leftarrow> return $ \<lparr>r_time = new_time, r_amount = r_amount refill_hd\<rparr>;
+     refills \<leftarrow> return (sc_refills sc);
+     refill_hd \<leftarrow> return (hd refills);
+
```

Comment:
I wasn't sure whether to truncate the refill here, as is done in the C, or just truncate to the head in each branch below